### ISSUE TITLE
`svdsolve` rrule error when not all values are converged

### ISSUE THREAD

**Issue by lkdvos** [Jutho/KrylovKit.jl#110]
In one of the PEPSKit testruns, we are running into an issue where the within the rrule for `svdsolve`, the eigensolver is failing to converge all requested values. In this case, the output `vals` and `vecs` are not of the expected length, leading to an out of bounds error.

The stacktrace is found [here](https://github.com/QuantumKitHub/PEPSKit.jl/actions/runs/12263340824/job/34214671852?pr=97)
The offending line is [this line](https://github.com/Jutho/KrylovKit.jl/blob/4d2a06f898fe1f87c8de240ed7c82c3591b9c934/ext/KrylovKitChainRulesCoreExt/svdsolve.jl#L237) .
There is already a warning being thrown, but I guess we have to limit `n` to the size of the computed vals and vecs.

**Comment by lkdvos** [comment]
Yeah, that's exactly what it's hitting. The log is a bit cryptic (blame Zygote), but this is relevant:
```julia
[ Info: CTMRG conv 4:	obj = +9.192864957180e-01	err = 4.1934293618e-11	time = 0.75 sec
┌ Warning: Invariant subspace of dimension 7 (up to requested tolerance `tol = 1.0e-8`), which is smaller than the number of requested eigenvalues (i.e. `howmany == 14`); setting `howmany = 7`.
└ @ KrylovKit ~/.julia/packages/KrylovKit/xccMN/src/eigsolve/arnoldi.jl:350
┌ Warning: `svdsolve` cotangent linear problem (7) returns unexpected result
└ @ KrylovKitChainRulesCoreExt ~/.julia/packages/KrylovKit/xccMN/ext/KrylovKitChainRulesCoreExt/svdsolve.jl:240
┌ Warning: `svdsolve` cotangent linear problem (14) returns unexpected result
└ @ KrylovKitChainRulesCoreExt ~/.julia/packages/KrylovKit/xccMN/ext/KrylovKitChainRulesCoreExt/svdsolve.jl:240
```

**Comment by Jutho** [comment]
Ok that is very weird. The different eigenvectors of the linear operator are supposed to be of the form

$$((\lambda_i - A)^{-1}\Delta_i , e_i)$$

with $e_i$ the length $n$ unit vector in the $i$ direction, and this for $i=1,...,n$. The initial vector is given by $(0, e_1 + e_2 + ... + e_n)$. That should have support in each of the eigenspaces.

I will try to reproduce this locally and investigate.

**Comment by lkdvos** [comment]
Apologies that I dont have a MWE, I am not sure how to construct something for which the forward algorithm passes but the reverse fails. In any case, my guess wasn't that there necessarily is something wrong with KrylovKit, I think our tolerances were just not configured properly, but I would expect KrylovKit to throw a warning and move on, rather than throwing a boundserror

**Comment by Jutho** [comment]
These are the singular values of the forward calculation:
```julia
[0.5827127960663261, 0.013616701381118779, 0.013616700815086135, 0.00031674568674968046, 7.429999268923436e-6, 7.4299980267290364e-6, 2.0722724010027094e-7, 2.0722722219200433e-7, 2.0722720896899213e-7, 2.0540682979946611e-7, 5.767011119369518e-9, 5.767010235714964e-9, 4.053212060637025e-9, 4.053210898281218e-9]
```

Clearly, there are degeneracies here, likely due to the SU2 symmetry of the Heisenberg model. As always, Krylov method cannot guarantee to find this; I think it is pretty amazing that it can find these degenerate values so well. But this is why the reverse algorithm chokes; there the degeneracies do cause the issue of the smaller invariant subspace.

I will need to think a bit about this. The fact that you don't get a "gauge dependence" warning, probably means that the adjoint variables associated with the singular vectors of degenerate singular values are all linearly dependent, so that maybe you can just replace it with having to solve a single linear system, and thus in the auxiliary eigenvalue problem that you build, you could throw out the degenerate copies.

**Comment by Jutho** [comment]
I assume this issue has not yet magically gone away yet. How pressing is this?

**Comment by lkdvos** [comment]
As far as I know we managed to work around this in PEPSKit and I have not seen it again myself.

**Comment by leburgel** [comment]
Ran into this again by accident. It doesn't occur in any of the regular tests anymore, but I can reproduce it by starting a PEPS optimization from a product state. Not really pressing I think, but very confusing to run into. MWE:

```julia
# using current PEPSKit.jl master...
# Pkg.add(;
#     url="https://github.com/QuantumKitHub/PEPSKit.jl",
#     rev="4de9e091b7f722933b28f097f1feaa8498ce5f04",
# )

using Random
using PEPSKit
using TensorKit
using OptimKit

g = 3.1
e = -1.6417 * 2
mˣ = 0.91

# initialize parameters
χbond = 2
χenv = 16
ctm_alg = SimultaneousCTMRG(; )
opt_alg = PEPSOptimize(;
    boundary_alg=ctm_alg, optimizer=LBFGS(4; gradtol=1e-3, verbosity=3)
)

# initialize states
H = transverse_field_ising(InfiniteSquare(); g)
Random.seed!(72534343814)
psi_init = product_peps(2, χbond; noise_amp=2e-1)

env_init = leading_boundary(CTMRGEnv(psi_init, ComplexSpace(χenv)), psi_init, ctm_alg)

# find fixedpoint
result = fixedpoint(psi_init, H, opt_alg, env_init)
```

```julia-repl
(dev) pkg> status
Status `~/git/PEPSKit.jl/dev/Project.toml`
  [0b1a1467] KrylovKit v0.9.3
⌅ [77e91f04] OptimKit v0.3.1
  [52969e89] PEPSKit v0.3.0 `https://github.com/QuantumKitHub/PEPSKit.jl#4de9e09`
  [07d1fe3e] TensorKit v0.14.3
  [9a3f8284] Random
```

```julia-repl
julia> versioninfo()
Julia Version 1.10.4
Commit 48d4fd48430 (2024-06-04 10:41 UTC)
Build Info:
  Official https://julialang.org/ release
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 16 × Intel(R) Core(TM) i9-10885H CPU @ 2.40GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-15.0.7 (ORCJIT, skylake)
Threads: 1 default, 0 interactive, 1 GC (on 16 virtual cores)
```

**Comment by leburgel** [comment]
This keeps popping up here and there, and becomes significantly worse when working with symmetries with small blocks. Do you think there is a sensible fix for this @Jutho, or does this really mean something is going fundamentally wrong in the simulations that we should safeguard against?

**Comment by Confusio** [comment]
After extensive investigation, I believe I have identified the source of the problem. The issue is not with retaining degenerate singular values per se, but rather with how they are truncated. In our case, when solving the Sylvester equation for $dP$

$$
C = dP \cdot C_{\rm keep} - C_{\rm discarded} \cdot dP,
$$

if the last singular value we keep is degenerate with the first one we discard, then both \(C_{\rm keep}\) and \(C_{\rm discarded}\) share the same eigenvalue. This eigenvalue degeneracy creates numerical difficulties when solving the Sylvester equation.

The solution is to ensure that during the truncation of CTMRG projections, all singular values that are degenerate with the truncation boundary are retained. A straightforward way to implement this is to modify the truncation function in `TensorKit.jl` as follows:

```julia
function _compute_truncdim(Σdata, trunc::TruncationDimension, p=2)
    I = keytype(Σdata)
    # Initialize truncation dimensions for each sector based on the available singular values.
    truncdim = SectorDict{I,Int}(c => length(v) for (c, v) in Σdata)
    S = []
    for (c, v) in Σdata
        # Repeat each singular value according to the degeneracy (given by dim(c)).
        for i in 1:dim(c)
            push!(S, v)
        end
    end
    # Sort the singular values in descending order.
    S = sort(vcat(S...); by = x -> -x)
    
    # Start with the requested truncation dimension.
    n = trunc.dim
    cutoff = eps(scalartype(S))
    # Define a threshold to decide if consecutive singular values are effectively degenerate.
    deg_threshold = 1.0 - sqrt(cutoff)
    n_above_cutoff = count(>=(cutoff), S / S[1])
    n = Int(min(n, n_above_cutoff))
    
    if n > 0 && n < length(S)
        # Extend the truncation point if the next singular value is nearly equal to the current one.
        last_index = n
        for i in n:(length(S) - 1)
            if S[i + 1] >= S[i] * deg_threshold
                last_index = i + 1
            else
                break
            end
        end
        n = last_index
    end
    
    # Adjust the truncation dimensions so that the total dimension does not exceed trunc.dim.
    while sum(dim(c) * d for (c, d) in truncdim) > trunc.dim
        cmin = _findnexttruncvalue(Σdata, truncdim, p)
        isnothing(cmin) && break
        truncdim[cmin] -= 1
    end
    
    return truncdim
end
```

Then, by setting the CTMRG projection with

```julia
SimultaneousCTMRG(; trscheme=truncdim(χenv))
```

the issue in my case was resolved.

This modification ensures that when the singular values around the truncation boundary are nearly equal (i.e. degenerate), they are all kept, thus avoiding the pitfalls in the Sylvester equation where identical eigenvalues in $C_{\rm keep}$ and $C_{\rm discarded}$ lead to numerical instability.

**Comment by Confusio** [comment]
> This keeps popping up here and there, and becomes significantly worse when working with symmetries with small blocks. Do you think there is a sensible fix for this [@Jutho](https://github.com/Jutho), or does this really mean something is going fundamentally wrong in the simulations that we should safeguard against?

As I explained above, "with symmetries with small blocks" will almost always cause the wrong truncation over degenerate singular values during the CTMRG truncation.

**Comment by Confusio** [comment]
I just fixed the issue by setting a small tolerance (1.0e-10) in Arnoldi algorithm occurred in my case.

**Comment by leburgel** [comment]
So it seems that at least in some cases, this issue is caused by the fact that the magnitude of the smallest remaining singular value lies below the tolerance of the `rrule_alg`, and this causes things to go wrong. A straightforward attempted fix would then be to determine the `rrule_alg` tolerance dynamically based on the magnitude of the smallest singular value (just `@set` it to the minimum of the provided value and the smallest singular value divided by 10 or something?). This might not fix all occurrences (degeneracies remain a mystery I guess), but at least it sounds like a sensible thing to do regardless, at least intuitively.

What do you think @Jutho?

**Comment by Jutho** [comment]
Yes; I do not feel great deviating from the arguments that were entered, but I guess there is (currently) no way to replicate such dynamic behavior in any other way. Is this still in the case where you actually use the KrylovKit `rrule` for a truncated SVD obtained from a full diagonalization?

**Comment by leburgel** [comment]
>  I do not feel great deviating from the arguments that were entered, but I guess there is (currently) no way to replicate such dynamic behavior in any other way.

I guess in spirit it's kind of a similar issue as when `svdsolve` is called with a `krylovdim` smaller than `howmany`, in the sense that the user-supplied arguments are just really not set up to even be able to solve the problem in a sensible way. I think now KrylovKit.jl just throws an error in this case, but in principle it could just as well use some heuristic to increase the `krylovdim` to an actually sensible value.

If you prefer to throw an error in case the `rrule` tolerance is larger than the magnitude of the smallest singular value to enforce that the arguments that were entered actually make sense, I think even that would be better than the kind of silent failures we're getting now.

**Comment by Jutho** [comment]
I am still a bit confused; given that you manually call KrylovKit's rrule, is it possible to set the tolerance after you already know `S` from within PEPSKit? I definitely agree that this will pop up elsewhere, although also the forward computation is ill-defined if singular values are smaller than tolerance.

**Comment by leburgel** [comment]
> I am still a bit confused; given that you manually call KrylovKit's rrule, is it possible to set the tolerance after you already know `S` from within PEPSKit? I definitely agree that this will pop up elsewhere, although also the forward computation is ill-defined if singular values are smaller than tolerance.

Yes, it's definitely possible to set the tolerance from within PEPSKit.jl. Part of the question (which may not have been clear from my explanation) is whether it would be better to circumvent this from PEPSKit.jl, or address it directly here. Since it seemed to me that this problem is more broad than just our use case I personally thought it might be good to address it here.

But indeed, while I keep kind of forgetting we're never going through the actual forward computation in our use case, it is definitely much more likely for the trouble to start in the forward computation in a normal use case. Given this, it does feel a bit strange to change to try to fix things in the backward pass when the forward was ill-defined. 

If not intervening, does it make sense to put some warnings in place? Performing a tolerance versus smallest singular value magnitude check in the rrule and throwing a warning if the former is larger is simple enough I guess and might be helpful. Will this issue be caught by any of the current warnings in the forward computation?
